### ISSUE TITLE
Enhancement: [max-params] allow not counting `this` at all

### ISSUE THREAD

**Issue by GerkinDev** [typescript-eslint/typescript-eslint#11574]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/max-params

### Description

Currently, it is only possible to ignore `this: void` parameter. However, when creating mixins or other functions bound to a type, `this: <type>` is useful, but is counted as a parameter, which it is not.

Adding an option `countThis` (similar to `countVoidThis`) with a default of `true` could allow skipping them while keeping backward compatibility

### Fail

```TypeScript
// With `{max: 2}`
function doSomething(this: MyType, param1: unknown, param2: unknown): unknown {}
```

### Pass

```TypeScript
// With `{max: 2, countThis: false}`
function doSomething(this: MyType, param1: unknown, param2: unknown): unknown {}
```

### Additional Info

I'm OK to PR

**Comment by bradzacher** [comment]
If you want to count the `this` parameter then you can just use the base rule.
Our extension rule solely exists to avoid counting the `this` parameter.

**Comment by Josh-Cena** [comment]
@bradzacher The base rule treats `this` as a parameter. The extension rule allows not treating `this: void` as a parameter. This issue is asking for an option to not treat `this` as a parameter in general (which I can clearly see the motivation for).

**Comment by bradzacher** [comment]
Oh.... I thought the entire point of the extension was to ignore the this parameter but I am wrong.

Regardless - we should be deprecating this rule - the upstream core rule now supports this parameters in the same way.

So we should reject this in favour of the core rule adopting the enhancement.

**Comment by Josh-Cena** [comment]
Oh right, this can be in core now.

**Comment by GerkinDev** [comment]
Indeed, I did not see that it was also in the core rule. Should I reopen this enhancement request there ? I suppose it was integrated by one of the members of this repo, so I wonder if you can take it from here or I should

**Comment by bradzacher** [comment]
> Should I reopen this enhancement request there

Yup - do that.
We're separate projects and they essentially merged our extension logic into their rules - there's little overlap between them.

If this is something you want to see - please file an issue with them.
